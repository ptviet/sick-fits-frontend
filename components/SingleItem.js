import React, { Component } from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import Head from 'next/head';
import Link from 'next/link';
import formatMoney from '../lib/formatMoney';
import ErrorMessage from './ErrorMessage';
import DeleteItem from './DeleteItem';
import AddToCart from './AddToCart';

const SingleItemStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${props => props.theme.bs};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
  .details {
    margin: 3rem;
    font-size: 2rem;
  }
`;

const ButtonList = styled.div`
  display: grid;
  width: 100%;
  border-top: 1px solid ${props => props.theme.lightgrey};
  grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));
  grid-gap: 1px;
  background: ${props => props.theme.lightgrey};
  & > * {
    background: white;
    border: 0;
    font-size: 1rem;
    padding: 1rem;
  }
`;

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      description
      price
      image
      largeImage
    }
  }
`;

class SingleItem extends Component {
  render() {
    return (
      <Query query={SINGLE_ITEM_QUERY} variables={{ id: this.props.id }}>
        {({ data, error, loading }) => {
          if (error) return <ErrorMessage error={error} />;
          if (loading) return <p>Loading...</p>;
          if (!data.item) return <p>No Item Found.</p>;
          const { item } = data;
          return (
            <>
              <ButtonList>
                <Link
                  href={{
                    pathname: '/update',
                    query: { id: item.id }
                  }}
                >
                  <a>Edit ✏️</a>
                </Link>
                <AddToCart id={item.id} />
                <DeleteItem id={item.id}>Delete ❌</DeleteItem>
              </ButtonList>
              <SingleItemStyles>
                <Head>
                  <title>Sick Fits | {item.title}</title>
                </Head>
                <img src={item.largeImage} alt={item.title} />
                <div className="details">
                  <h2>{item.title}</h2>
                  <h4>{formatMoney(item.price)}</h4>
                  <p>{item.description}</p>
                </div>
              </SingleItemStyles>
            </>
          );
        }}
      </Query>
    );
  }
}

export default SingleItem;
export { SINGLE_ITEM_QUERY };
