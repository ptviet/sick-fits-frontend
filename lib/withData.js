import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpointProd } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    // uri: process.env.NODE_ENV === 'development' ? endpoint : endpointProd,
    uri: 'https://sp-sick-fits-yoga-prod.herokuapp.com',
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include'
        },
        headers
      });
    },
    // Local data
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            // Read the cartOpen value from the cache
            const { cartOpen } = cache.readQuery({ query: LOCAL_STATE_QUERY });
            // Write new cartOpen state
            const data = {
              data: {
                cartOpen: !cartOpen
              }
            };
            cache.writeData(data);
            return data;
          }
        }
      },
      defaults: {
        cartOpen: false
      }
    }
  });
}

export default withApollo(createClient);
