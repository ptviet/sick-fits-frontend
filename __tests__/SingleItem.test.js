import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';
import SingleItem, { SINGLE_ITEM_QUERY } from '../components/SingleItem';
import { fakeItem } from '../lib/testUtils';

describe('<SingleItem />', () => {
  let wrapper;
  let mocks;

  it('should render with proper data', async () => {
    mocks = [
      {
        // when
        request: { query: SINGLE_ITEM_QUERY, variables: { id: fakeItem().id } },
        // then
        result: { data: { item: fakeItem() } }
      }
    ];

    wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id={fakeItem().id} />
      </MockedProvider>
    );

    expect(wrapper.text()).toContain('Loading...');

    await wait();
    wrapper.update();

    expect(toJson(wrapper.find('h2'))).toMatchSnapshot();
    expect(toJson(wrapper.find('img'))).toMatchSnapshot();
    expect(toJson(wrapper.find('p'))).toMatchSnapshot();
  });

  it('should show an error with a not found item', async () => {
    mocks = [
      {
        // when
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        // then
        result: { errors: [{ message: 'Item not found!' }] }
      }
    ];

    wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>
    );

    await wait();
    wrapper.update();

    const item = wrapper.find('[data-test="graphql-error"]');

    expect(item.text()).toContain('Item not found!');
    expect(toJson(item)).toMatchSnapshot();
  });
});
