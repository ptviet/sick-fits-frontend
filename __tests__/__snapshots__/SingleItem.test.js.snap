// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem /> should render with proper data 1`] = `
<h2>
  dogs are best
</h2>
`;

exports[`<SingleItem /> should render with proper data 2`] = `
<img
  alt="dogs are best"
  src="dog.jpg"
/>
`;

exports[`<SingleItem /> should render with proper data 3`] = `
<p>
  dogs
</p>
`;

exports[`<SingleItem /> should show an error with a not found item 1`] = `
<p
  data-test="graphql-error"
>
  Item not found!
</p>
`;
